v3/mathjax3-ts/util/LinkedList.ts

```diff
@@ -0,0 +1,356 @@
+        item.next.prev = this.list;
+        item.next = item.prev = null;
+        return item.data as DataClass;
+    }
+
+    /*
+     * Empty the list
+     *
+     * @return{LinkedList}  The LinkedList object (for chaining)
+     */
+    public clear() {
+        this.list.next = this.list.prev = this.list;
+        return this;
+    }
+
+    /*
+     * Make the list iterable and return the data from the items in the list
+     *
+     * @return{{next: Function}}  The object containing the iterator's next() function
+     */
+    public [Symbol.iterator](): Iterator<DataClass> {
+        let current = this.list;
+        return {
+            next() {
+                current = current.next;
+                return (current.data === END ?
+                        {value: null, done: true} :
+                        {value: current.data, done: false}) as IteratorResult<DataClass>;
+            }
+        };
+    }
+
+    /*
+     * An iterator for the list in reverse order
+     *
+     * @return{Object}  The iterator for walking the list in reverse
+     */
+    public reversed() {
+        let current = this.list;
+        return {
+            [Symbol.iterator](): Iterator<DataClass> {
+                return this;
+            },
+            next() {
+                current = current.prev;
+                return (current.data === END ?
+                        {value: null, done: true} :
+                        {value: current.data, done: false}) as IteratorResult<DataClass>;
+            },
+            toArray() {
+                return Array.from(this) as DataClass[];
+            }
+        };
+    }
+
+    /*
+     * Insert a new item into a sorted list in the correct locations
+     *
+     * @param{DataClass} data   The data item to add
+     * @param{SortFn} isBefore  The function used to order the data
+     * @param{LinkedList}       The LinkedList object (for chaining)
+     */
+    public insert(data: DataClass, isBefore: SortFn<DataClass> = null) {
+        if (isBefore === null) {
+            isBefore = this.isBefore.bind(this);
+        }
+        let item = new ListItem<DataClass>(data);
+        let cur = this.list.next;
+        while (cur.data !== END && isBefore(cur.data as DataClass, item.data as DataClass)) {
+            cur = cur.next;
+        }
+        item.prev = cur.prev;
+        item.next = cur;
+        cur.prev.next = cur.prev = item;
+        return this;
+    }
+
+    /*
+     * Sort the list using an optional sort function
+     *
+     * @param{SortFn} isBefore  The function used to order the data
+     * @return{LinkedList}      The LinkedList object (for chaining)
+     */
+    public sort(isBefore: SortFn<DataClass> = null) {
+        if (isBefore === null) {
+            isBefore = this.isBefore.bind(this);
+        }
+        //
+        //  Make an array of singleton lists
+        //
+        let lists: LinkedList<DataClass>[] = [];
+        for (const item of this) {
```

Comment:
Yes, that is the purpose of the `[Symbol.iterator]()` method above.  This is what required Typescript 2.3, since earlier versions didn't implement it for the ES5 target (but did for ES6 target).